lib/OpenQA/CacheService.pm

```diff
@@ -30,7 +31,25 @@ sub startup {
     $self->defaults(appname => 'openQA Cache Service');
     # Provide help to users early to prevent failing later on
     # misconfigurations
-    return if $ENV{MOJO_HELP};
+    return $self->{_app_return_code} = 0 if $ENV{MOJO_HELP};
+
+    # Stop the service after a critical database error
+    $self->helper(
+        'reply.exception' => sub {
+            my ($c, $error) = @_;
+
+            my $log = $c->app->log;
+            $error = Mojo::Exception->new($error) unless blessed $error && $error->isa('Mojo::Exception');
+            $error = $error->inspect->to_string;
+            chomp $error;
+            $log->error($error);
+            $c->render(text => $error, status => 500);
+            if ($error =~ qr/(database disk image is malformed|no such (table|column))/) {
+                $c->app->{_app_return_code} = 1;    # ensure the return code is non-zero
+                $log->error('Stopping service after critical database error');
+                Mojo::IOLoop->singleton->stop_gracefully;
```

Comment:
This only stops the current process, which is usually just a pre-fork worker process, not the whole web server.

Reply:
Is there a way to stop the server manager?

Reply:
Only signals. The list is [here](https://metacpan.org/pod/Mojo::Server::Prefork#MANAGER-SIGNALS).